### PR TITLE
PP-1132, PP-1133, PP-1134, PP-1135, PP-1136, PP-1126: Windows build scripts enhancement

### PR DESCRIPTION
**Pull request by hirenvadalia**
#### Issue-ID

* **[PP-1132](https://pbspro.atlassian.net/browse/PP-1132)**
* **[PP-1133](https://pbspro.atlassian.net/browse/PP-1133)**
* **[PP-1134](https://pbspro.atlassian.net/browse/PP-1134)**
* **[PP-1135](https://pbspro.atlassian.net/browse/PP-1135)**
* **[PP-1136](https://pbspro.atlassian.net/browse/PP-1136)**
* **[PP-1126](https://pbspro.atlassian.net/browse/PP-1126)**

#### Problem description
* Windows build broken in debug configuration
* CI doesn't do debug build
* Windows installer needs cleanup for more readability
* No MSI available for debug build
* MSI reports wrong PBS version

#### Cause / Analysis
* None

#### Solution description
* Changed windows build scripts to compile all PBS deps in debug mode for PBS debug configuration
* Fixed windows debug build and link with debug version of PBS deps instead release version
* Added debug build in matrix for AppVeyor CI, also added support for AppVeyor CI build cache to reduce time of CI
* Cleaned up windows installer code
* Added support to generate MSI in debug configuration
* Fixed wrong PBS version in MSI by added PBS version while generating MSI

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by bremanandjk** [comment]
Hi @hirenvadalia,

Thanks for making this changes, Can you post the test results for building and installing the msi. (which includes registering and starting the services; tests may extend to submit jobs and showing running status of jobs). If there are any changes in packaging, share that steps too. Thanks

I have followed below steps to build msi, its fine at building, but was facing issue at installation. 
a) Compiled the source code using MS visual studio.
b) Ran the prep.bat script like "prep.bat C:\Users\Brem\Desktop\H_pbs_win_installer\binaries"
c) Ran the create_msi.bat to generate the msi, Its successfully created the msi at win_build directory. 
d) Installed the newly created msi, 
e) I am facing failure message while running "win_postinstall.py" at installing Visual C++

C:\Program Files (x86)\PBS\exec\etc>win_postinstall.py -u username -p password
Post installation process started
Validating given username and password
Successfully validated given username and password
Installing Visual C++ redistributable
Failed to install Visual C++ redistributable

Let me know do i need to do some additional steps to generate and install OSS PBS Pro. Thanks

**Comment by hirenvadalia** [comment]
@bremanandjk Thanks for trying out... I am still collecting all test results and will post once all are ready... 
Now regarding failure: Can you please provide me following details for debugging?
1. exit code of win_postinstall.py
2. output of `set` command in terminal in which you ran win_postinstall.py
2. output of `dir` command in directory `C:\Users\Brem\Desktop\H_pbs_win_installer\binaries`

**Comment by bremanandjk** [comment]
Hi @hirenvadalia,

Thanks for looking into it, 
1) Exit code of win_postinstall.py is "6"

```
C:\Program Files (x86)\PBS\exec\etc>win_postinstall.py -u username -p password
Post installation process started
Validating given username and password
Successfully validated given username and password
Installing Visual C++ redistributable
Failed to install Visual C++ redistributable

C:\Program Files (x86)\PBS\exec\etc>echo %ERRORLEVEL%
6
```
2) output of set command in terminal in which you ran win_postinstall.py
```
C:\Program Files (x86)\PBS\exec\etc>set
ALLUSERSPROFILE=C:\ProgramData
APPDATA=C:\Users\Brem\AppData\Roaming
CommonProgramFiles=C:\Program Files\Common Files
CommonProgramFiles(x86)=C:\Program Files (x86)\Common Files
CommonProgramW6432=C:\Program Files\Common Files
COMPUTERNAME=SPARK
ComSpec=C:\Windows\system32\cmd.exe
FP_NO_HOST_CHECK=NO
HOMEDRIVE=C:
HOMEPATH=\Users\Brem
LOCALAPPDATA=C:\Users\Brem\AppData\Local
LOGONSERVER=\\SPARK
NUMBER_OF_PROCESSORS=4
OS=Windows_NT
Path=C:\Python27\;C:\Python27\Scripts;C:\ProgramData\Oracle\Java\javapath;C:\Win
dows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPow
erShell\v1.0\;C:\Program Files\Perforce;C:\Program Files\Perforce\DVCS\;C:\Progr
am Files (x86)\WiX Toolset v3.11\bin;C:\Program Files (x86)\Meld;C:\Program File
s (x86)\PBS\exec\bin;C:\Program Files (x86)\PBS\exec\sbin;C:\Program Files (x86)
\PBS\exec\lib;C:\Program Files (x86)\PBS\exec\pgsql\bin;C:\Program Files (x86)\P
BS\exec\python;C:\Program Files (x86)\PBS\exec\python\Scripts;C:\Program Files (
x86)\PBS\exec\python\Lib\site-packages\pywin32_system32
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
PBS_CONF_FILE=C:\Program Files (x86)\PBS\pbs.conf
PBS_EXEC=C:\Program Files (x86)\PBS\exec\
PROCESSOR_ARCHITECTURE=AMD64
PROCESSOR_IDENTIFIER=Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
PROCESSOR_LEVEL=6
PROCESSOR_REVISION=5e03
ProgramData=C:\ProgramData
ProgramFiles=C:\Program Files
ProgramFiles(x86)=C:\Program Files (x86)
ProgramW6432=C:\Program Files
PROMPT=$P$G
PSModulePath=C:\Windows\system32\WindowsPowerShell\v1.0\Modules\
PUBLIC=C:\Users\Public
SESSIONNAME=Console
SystemDrive=C:
SystemRoot=C:\Windows
TEMP=C:\Users\Brem\AppData\Local\Temp
TMP=C:\Users\Brem\AppData\Local\Temp
USERDOMAIN=SPARK
USERNAME=Brem
USERPROFILE=C:\Users\Brem
VS140COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\Tools
\
VS90COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio 9.0\Common7\Tools\
windir=C:\Windows
windows_tracing_flags=3
windows_tracing_logfile=C:\BVTBin\Tests\installpackage\csilogfile.log
WIX=C:\Program Files (x86)\WiX Toolset v3.11\

C:\Program Files (x86)\PBS\exec\etc>
```
3) output of dir command in directory C:\Users\Brem\Desktop\H_pbs_win_installer\binaries

```
C:\Users\Brem\Desktop\H_pbs_win_installer\binaries>dir
 Volume in drive C has no label.
 Volume Serial Number is 6249-4000

 Directory of C:\Users\Brem\Desktop\H_pbs_win_installer\binaries

12/06/2017  11:12 AM    <DIR>          .
12/06/2017  11:12 AM    <DIR>          ..
12/06/2017  11:11 AM    <DIR>          alsdk
12/06/2017  11:11 AM    <DIR>          libedit
12/06/2017  11:11 AM    <DIR>          libical
12/06/2017  11:11 AM    <DIR>          openssl
12/06/2017  11:12 AM    <DIR>          pgsql
12/06/2017  11:12 AM    <DIR>          python
12/06/2017  11:12 AM    <DIR>          python_x64
12/06/2017  11:12 AM    <DIR>          swig
12/06/2017  11:12 AM    <DIR>          tcltk
               0 File(s)              0 bytes
              11 Dir(s)  35,139,477,504 bytes free

C:\Users\Brem\Desktop\H_pbs_win_installer\binaries>
```

Dir for "etc" also provided for debugging.

```
C:\Program Files (x86)\PBS\exec\etc>dir
 Volume in drive C has no label.
 Volume Serial Number is 6249-4000

 Directory of C:\Program Files (x86)\PBS\exec\etc

12/06/2017  11:56 AM    <DIR>          .
12/06/2017  11:56 AM    <DIR>          ..
12/06/2017  11:27 AM                14 createdb.bat
12/06/2017  11:27 AM                14 create_svr_defaults.bat
12/06/2017  11:13 AM            10,522 pbs_db_schema.sql
12/06/2017  11:13 AM             1,774 pbs_dedicated
12/06/2017  11:13 AM             2,612 pbs_holidays
12/06/2017  11:13 AM             2,643 pbs_holidays.2017
12/06/2017  11:13 AM             1,891 pbs_resource_group
12/06/2017  11:13 AM            16,498 pbs_sched_config
06/19/2017  10:09 AM         5,073,240 vcredist_x86.exe
12/06/2017  11:20 AM            14,208 win_postinstall.py
              10 File(s)      5,123,416 bytes
               2 Dir(s)  35,139,477,504 bytes free

C:\Program Files (x86)\PBS\exec\etc>
```

**Comment by hirenvadalia** [comment]
@bremanandjk I tried multiple time from start to end but I am not able to reproduce issue. For me everything is working fine. Can you please try with clean environment and see whether you are facing issue or not?

Here is logs from my tests:
```
C:\Users\pbsbuild>"C:\Program Files (x86)\PBS\exec\python\python.exe" "C:\Program Files (x86)\PBS\exec\etc\win_postinstall.py" -u pbsbuild -p *********
Post installation process started
Validating given username and password
Successfully validated given username and password
Installing Visual C++ redistributable
Successfully installed Visual C++ redistributable
Creating PBSPro configuration at C:\Program Files (x86)\PBS\pbs.conf
Successfully created PBSPro configuration
Creating PBSPro home
Initializing PBSPro database
Successfully initialized PBSPro database
Successfully registered service PBS_SCHED
Successfully started PBS_SCHED service
Successfully registered service PBS_COMM
Successfully started PBS_COMM service
Successfully registered service PBS_MOM
Successfully started PBS_MOM service
Successfully registered service PBS_SERVER
Successfully started PBS_SERVER service
Setting PBSPro default configuration
Successfully configured PBSPro defaults
Successfully completed post installation process

C:\Users\pbsbuild>
```
```
C:\Users\pbsbuild>qstat

C:\Users\pbsbuild>qmgr -c "s s acl_roots+=pbsbuild"

C:\Users\pbsbuild>pbs_password
Enter user's password:
Verifying - Enter user's password:
Changed user pbsbuild's PBS password on server winbuild

C:\Users\pbsbuild>qsub -- pbs-sleep 1
0.winbuild

C:\Users\pbsbuild>qstat -f
Job Id: 0.winbuild
    Job_Name = STDIN
    Job_Owner = pbsbuild@winbuild
    resources_used.cput = 00:00:00
    resources_used.mem = 0kb
    job_state = E
    queue = workq
    server = winbuild
    Checkpoint = u
    ctime = Thu Dec 07 10:20:37 2017
    Error_Path = winbuild:C:/Users/pbsbuild/STDIN.e0
    exec_host = winbuild/0
    exec_vnode = (winbuild:ncpus=1)
    Hold_Types = n
    Join_Path = n
    Keep_Files = n
    Mail_Points = a
    mtime = Thu Dec 07 10:20:38 2017
    Output_Path = winbuild:C:/Users/pbsbuild/STDIN.o0
    Priority = 0
    qtime = Thu Dec 07 10:20:37 2017
    Rerunable = True
    Resource_List.ncpus = 1
    Resource_List.nodect = 1
    Resource_List.place = pack
    Resource_List.select = 1:ncpus=1
    schedselect = 1:ncpus=1
    stime = Thu Dec 07 10:20:37 2017
    session_id = 2584
    jobdir = C:/Users/pbsbuild/Documents/PBS Pro
    substate = 51
    Variable_List = PBS_O_HOME=/,
        PBS_O_PATH=C:/Program Files (x86)/WiX Toolset v3.11/bin;C:/Tools;C:/Wi
        ndows/system32;C:/Windows;C:/Windows/System32/Wbem;C:/Windows/System32/
        WindowsPowerShell/v1.0/;C:/Program Files/sysinternalssuite;C:/Program F
        iles/TortoiseSVN/bin;C:/Strawberry/c/bin;C:/Strawberry/perl/site/bin;C:
        /Strawberry/perl/bin;C:/Program Files/CMake/bin;C:/Program Files (x86)/
        PBS/exec/bin;C:/Program Files (x86)/PBS/exec/sbin;C:/Program Files (x86
        )/PBS/exec/lib;C:/Program Files (x86)/PBS/exec/pgsql/bin;C:/Program Fil
        es (x86)/PBS/exec/python;C:/Program Files (x86)/PBS/exec/python/Scripts
        ;C:/Program Files (x86)/PBS/exec/python/Lib/site-packages/pywin32_syste
        m32,PBS_O_WORKDIR=C:/Users/pbsbuild,PBS_O_SYSTEM=VER_PLATFORM_WIN32_NT,
        PBS_O_QUEUE=workq,PBS_O_HOST=winbuild
    euser = pbsbuild
    egroup = Administrators
    hashname = 0.winbuild
    queue_rank = 793749058
    queue_type = E
    comment = Job run at Thu Dec 07 at 10:20 on (winbuild:ncpus=1)
    alt_id = HomeDirectory=C:\Users\pbsbuild\Documents\PBS Pro
    etime = Thu Dec 07 10:20:37 2017
    run_count = 1
    Exit_status = 0
    Submit_arguments = -- pbs-sleep 1
    executable = <jsdl-hpcpa:Executable>pbs-sleep</jsdl-hpcpa:Executable>
    argument_list = <jsdl-hpcpa:Argument>1</jsdl-hpcpa:Argument>
    project = _pbs_project_default
    run_version = 1
```
```
C:\Users\pbsbuild>tracejob 0

Job: 0.winbuild

12/07/2017 10:20:37  L    Considering job to run
12/07/2017 10:20:37  A    user=pbsbuild group=Administrators
                          project=_pbs_project_default jobname=STDIN queue=workq

                          ctime=1512622237 qtime=1512622237 etime=1512622237
                          start=1512622237 exec_host=winbuild/0
                          exec_vnode=(winbuild:ncpus=1) Resource_List.ncpus=1
                          Resource_List.nodect=1 Resource_List.place=pack
                          Resource_List.select=1:ncpus=1
                          resource_assigned.ncpus=1
12/07/2017 10:20:37  S    Job Queued at request of pbsbuild@winbuild, owner =
                          pbsbuild@winbuild, job name = STDIN, queue = workq
12/07/2017 10:20:37  S    Job Run at request of Scheduler@winbuild on exec_vnode

                          (winbuild:ncpus=1)
12/07/2017 10:20:37  L    Job run
12/07/2017 10:20:37  A    queue=workq
12/07/2017 10:20:37  M    allowed pbsbuild to access window station and desktop,

                          User pbsbuild passworded
12/07/2017 10:20:37  M    Started, pid = 2584
12/07/2017 10:20:37  M    Started, pid = 2584
12/07/2017 10:20:38  S    Obit received momhop:1 serverhop:1 state:4 substate:42

12/07/2017 10:20:38  M    task 1 terminated
12/07/2017 10:20:38  M    Terminated
12/07/2017 10:20:38  M    task 00000001 cput= 0:00:00
12/07/2017 10:20:38  M    kill_job
12/07/2017 10:20:38  M    winbuild cput= 0:00:00 mem=0kb
12/07/2017 10:20:38  M    Obit sent
12/07/2017 10:20:38  M    copy file cred request received
12/07/2017 10:20:38  M    User pbsbuild passworded
12/07/2017 10:20:39  M    staged 2 items out over 0:00:01
12/07/2017 10:20:49  M    post_cpyfile: entered 0.winbuild
12/07/2017 10:20:49  M    post_cpyfile: done 0.winbuild
12/07/2017 10:20:49  A    user=pbsbuild group=Administrators
                          project=_pbs_project_default jobname=STDIN queue=workq

                          ctime=1512622237 qtime=1512622237 etime=1512622237
                          start=1512622237 exec_host=winbuild/0
                          exec_vnode=(winbuild:ncpus=1) Resource_List.ncpus=1
                          Resource_List.nodect=1 Resource_List.place=pack
                          Resource_List.select=1:ncpus=1 session=2584
                          alt_id="HomeDirectory=C:\Users\pbsbuild\Documents\PBS
                          Pro" end=1512622249 Exit_status=0
                          resources_used.cput=00:00:00 resources_used.mem=0kb
                          run_count=1
12/07/2017 10:20:50  S    Exit_status=0 resources_used.cput=00:00:00
                          resources_used.mem=0kb
12/07/2017 10:20:50  M    delete job request received
12/07/2017 10:20:50  M    kill_job

C:\Users\pbsbuild>
```

```
C:\Users\pbsbuild>dir
 Volume in drive C has no label.
 Volume Serial Number is 90E4-A6BE

 Directory of C:\Users\pbsbuild

12/07/2017  10:20 AM    <DIR>          .
12/07/2017  10:20 AM    <DIR>          ..
08/10/2017  09:44 PM    <DIR>          .cpanm
11/30/2017  03:12 PM                 9 .pbs_qmgr_history
12/07/2017  10:05 AM    <DIR>          builds
11/23/2017  04:55 PM    <DIR>          Contacts
11/30/2017  12:02 PM    <DIR>          Desktop
11/24/2017  10:21 PM    <DIR>          Documents
11/23/2017  04:55 PM    <DIR>          Downloads
11/23/2017  04:55 PM    <DIR>          Favorites
11/23/2017  04:55 PM    <DIR>          Links
11/23/2017  04:55 PM    <DIR>          Music
11/23/2017  04:55 PM    <DIR>          Pictures
11/23/2017  04:55 PM    <DIR>          Saved Games
11/23/2017  04:55 PM    <DIR>          Searches
12/07/2017  10:20 AM                 0 STDIN.e0
12/07/2017  10:20 AM                 0 STDIN.o0
11/23/2017  04:55 PM    <DIR>          Videos
               3 File(s)              9 bytes
              15 Dir(s)  172,059,467,776 bytes free

C:\Users\pbsbuild>
```

**Comment by bremanandjk** [comment]
Hi @hirenvadalia ,

Thanks for trying, I am facing the same issue again on my machine, in order to confirm that it is not related to the environment, I have tried to install today's mainline msi package, Its working fine for me in this machine. 

```
C:\Program Files (x86)\PBS\exec\etc>"C:\Program Files (x86)\PBS\exec\python\pyth
on.exe" win_postinstall.py -u Brem -p login123
Post install process started...
This script may take a few minutes to run...
User Validation Success
pbs.conf file created
Installing Visual C++ Redist..
Installation done..
Executing createdbfile as install user
Database initialization successful
postgresql file Successfully configured
pg_hba file Successfully configured
InstallDataService completed
securing C:/PROGRA~2/PBS for read access by Everyone
securing C:/PROGRA~2/PBS/PBS~1.CON for read access by Everyone
securing C:/PROGRA~2/PBS/exec/ for read access by Everyone
Securing and recreating pbs_home directories
Registering and starting PBS services...
Scheduler service registered
Comm service registered
Server service registered
PBS Server database created...
Executing serverdbfile as installing user
Server defaults configs are done
Mom service registered
MoM started
Successfully installed

C:\Program Files (x86)\PBS\exec\etc>
```

I could think of two potential cause for this failure 

1) The PBS directory permissions are read-only for everyone after GUI installation. (In mainline, we used to set the required permissions as a part installer). Maybe we could try executing the permission set line   [no 162](https://github.com/PBSPro/pbspro/pull/479/files#diff-46b2952195898217641c4395b2187cc3R162) at first place.
2) we could try using "/passive" & "/norestart" 

P.S : As part of this PR, we are modifying the interfaces too, it good to update respective [EDD](https://pbspro.atlassian.net/wiki/spaces/PD/pages/53188191/PP-854+PBS+Pro+for+Windows+needs+an+installer)

**Comment by hirenvadalia** [comment]
@bremanandjk Thanks. But I don't think its related to permission. As we show in your machine whatever problem we are facing is because newer version of VCRedist is already installed so while win_postinstall tries to install VCRedist from PBS_EXEC\etc directory it fails since it is older version...

Anyways, now I got root cause of problem so I will look into it and will update code accordingly...

Also can you please point which interface is changed? because I tried my best to not to change any interface as I am just cleaning up installer code only...

**Comment by bremanandjk** [comment]
@hirenvadalia Thanks for looking into this root cause. 

I meant "interface changes" as console output logs. I can see change in that

Example: Previously we used to print `"User Validation Success" `but now it is 
```
"Validating given username and password
Successfully validated given username and password"
```
These installation messages are good to document in EDD for debugging purpose.

And, 

Previously we have documented the "Exit code" with respective error, that too modified in this PR

Example. VCRedist installation failure's exit code was "5" , now its modified to "6" 

Hope you got the modification details. Let me know if further details needed.

**Comment by bremanandjk** [comment]
As we discussed, I am updating the current understanding of our discussion. 

1) In mainline, we are installing the "vcredist_x86.exe" using "os.system()" with "**/passive** /norestart" flags, It would install only it is feasible to install. If in case of higher version already installed in the system, it will ignore and be passive. Thus it didn't affect the installation process. (anyhow we have higher version to work on it)

2) But in the **new** approach, we are calling subprocess.Popen() module. here, passing "**/q** /norestart" flags. if in the case there is already the latest version installed, it would return code "5100".

I like the approach of comparing the error code to check any higher version is already installed and take a decision. if required/desired version is already available, we should proceed the installation process instead of failing to proceed on.

File: src/cmds/scripts/win_postinstall.py [hunk 1]
```diff
@@ -1,633 +1,406 @@
...
-    position_of_exec = install_path.find(exec_etc_path)
-    if position_of_exec:
-        pbs_path = install_path[:position_of_exec]
-    else:
-        print "Error while finding the PBS Pro installation path"
+        else:
+            print 'Unreocgnized option %s' % o
+            usage()
+            sys.exit(3)
+    if username is None or password is None:
```
Comment (sid2364) [inline src/cmds/scripts/win_postinstall.py]: Both username and password are declared to '' but the check here is for None.

Comment (hirenvadalia) [reply src/cmds/scripts/win_postinstall.py]: @sid2364 Thanks for catching. I have addressed you comment please have a look into latest commit.

Comment (subhasisb) [reply src/cmds/scripts/win_postinstall.py]: At some point we should make the postinstall program the same one for all platforms. Ie. for Linux too. Could you file a ticket for the future?

File: src/cmds/scripts/win_postinstall.py [hunk 2]
```diff
@@ -1,633 +1,406 @@
...
+    create_home()
+    init_db(username, password)
+    __run_cmd([os.path.join(pbs_bin, 'pbs_mkdirs.exe')])
+    register_and_start_services(username, password)
+    set_svr_defaults()
+    print 'Successfully completed post installation process'
+
+
+if __name__ == '__main__':
+    main()
```
Comment (subhasisb) [inline src/cmds/scripts/win_postinstall.py]: All the print's accompanying the exits seem to be on stdout. They should really be on stderr.

File: src/cmds/scripts/win_postinstall.py (hunk 1 above)
Comment (hirenvadalia) [reply src/cmds/scripts/win_postinstall.py]: @subhasisb Good idea! Created [PP-1182](https://pbspro.atlassian.net/browse/PP-1182)

File: .appveyor/_build_python_32bit.bat [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+@echo on
+setlocal
+
+call "%~dp0set_paths.bat" %~1
+
+cd "%BINARIESDIR%"
+
+if not defined PYTHON_VERSION (
+    echo "Please set PYTHON_VERSION to Python version!"
+    exit /b 1
```
[the remark this answers is not in the document]
Comment (hirenvadalia) [reply .appveyor/_build_python_32bit.bat]: @subhasisb May be I am missing something but at the end of script does not do cd back to originating directory so I think we don't need to change anything for this comment.

File: src/cmds/scripts/win_postinstall.py (hunk 2 above)
Comment (hirenvadalia) [reply src/cmds/scripts/win_postinstall.py]: @subhasisb ok. will change those to stderr in next commit.

File: src/cmds/scripts/win_postinstall.py (hunk 1 above)
Comment (subhasisb) [reply src/cmds/scripts/win_postinstall.py]: Thanks for filing the issue

File: .appveyor/_build_python_32bit.bat (hunk 3 above)
[the remark this answers is not in the document]
Comment (subhasisb) [reply .appveyor/_build_python_32bit.bat]: ok looks like the parent scripts take care of that, so fine with me.

**Comment by subhasisb** [comment]
Looks good to me Hiren. Please rebase and squash.

**Comment by subhasisb** [comment]
@hirenvadalia seems like there is one failed test case on Debian. Please check

**Comment by mike0042** [comment]
There is one test that fails occasionally on Debian. There is already a ticket open for the same test failing in the nightly builds. I restarted the Travis build for Debian. There is still a Codacy issue to address.

**Comment by subhasisb** [comment]
Yes, I agree the subprocess.Popen flagging by codacy is actually a codacy problem. We should turn that off.

**Comment by subhasisb** [comment]
I will go ahead and merge

**Comment by varunsonkar** [comment]
@hirenvadalia ,
for subprocess.Popen "shell=False" is by default, you may not need to pass it explicitely. May be codacy is catching because of that. Try without passing shell=False.

**Comment by hirenvadalia** [comment]
@varunsonkar Thanks. But codacy explicitly looked for 'shell=True' irrespective you as added shell arg or not (as per codacy source code). Anyways, in my first version it was without shell argument but codacy was throwing some different error so I added shell argument now codacy is giving false issue, which we have to ignore.

**Pull request merged**